### PR TITLE
[qunit] use asserts for strict equality assertions

### PR DESCRIPTION
**Pull request by BarryThePenguin**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://api.qunitjs.com/assert/
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

I'm not entirely sure this is correct.. but I'll see what others think!

[`qunit` assertions](https://api.qunitjs.com/assert/) collect errors from tests, and report them at the end of a test. They don't actually throw an error during the test when an assertion fails

However, it would still be helpful to have the ergonomics of `asserts` when using `qunit`. Something along the lines of..

```ts
type Foo = {
  bar?: string
}

QUnit.test('example', (assert: Assert) {
  const foo: Foo | null = fooList.first()
  assert.ok(foo) // asserts foo exists
  assert.strictEqual(foo.bar, 'baz') // won't complain that foo is possibly 'null'
})
```

I've based this work off #42786

### CODE REVIEW

**Comment by BarryThePenguin** [comment]
I understand that this change impacts other DefinitelyTyped packages, so I'll wait for some input before updating the affected packages

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ember__test-helpers ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember__test-helpers@latest/))
was missing the following properties:
1. name
1. init
1. included
1. treeForAddonTestSupport
1. treeForVendor
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 94cc2cccea444c1eab4f3e358d5ca4d3d2e4e9fb
</p>

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Thanks for opening this! I'm marking it as "Request Changes" *not* because I dislike the changes in principle—I've wanted them myself multiple times lately!—but because I'm torn on the annoyance factor this would introduce for *every* user of the QUnit types.

The problem is that (as you discovered and updated all the tests to account for), adding this functionality makes it so that *anywhere* you use `.ok`, `.notOk`, or `.strictEqual`, they *have* to add `assert: Assert` to their usage, which also means they have to *import* it. This is a massive change that will simply break everyone's existing tests, and I'm very unsure it's worth it. 😕 The workarounds which exist (particularly the use of the non-null assertion `!` operator) are not great, and if we had a way to actually signal a breaking change on DT, I'd be all in favor… but as it stands, we *don't* have any mechanic for signaling that, and people are apt to end up hitting this in what they perceive to be a minor version bump. 😬 

I'd really like input from @dfreeman and @jamescdavis (and any others) before we think about landing this.

**Comment by BarryThePenguin** [comment]
> Thanks for opening this! I'm marking it as "Request Changes" _not_ because I dislike the changes in principle—I've wanted them myself multiple times lately!—but because I'm torn on the annoyance factor this would introduce for _every_ user of the QUnit types.

Agreed! I opened this with the hope that a discussion would figure out if this is possible or not.

> they _have_ to add `assert: Assert` to their usage, which also means they have to _import_ it.

I thought the QUnit `Assert` type was global? That's how I've been using it. I haven't had to import it? It would still need to be added, yes 

> This is a massive change that will simply break everyone's existing tests, and I'm very unsure it's worth it. 😕

Ok, that's understandable..

Another issue with using `asserts` is, QUnit doesn't actually throw an assertions error when an assertion fails. It collects assertion messages and shows those messages after the test has run. In my previous example..

```ts
type Foo = {
  bar?: string
}

QUnit.test('example', (assert: Assert) {
  const foo: Foo | null = fooList.first()
  assert.ok(foo) // asserts foo exists
  assert.strictEqual(foo.bar, 'baz') // won't complain that foo is possibly 'null'
})
```

..if `fooList.first()` returns `null`, the `strictEqual` line would fail with a `TypeError`

> Cannot read property 'bar' of null

So there's still a bit of ambiguity there where TypeScript would say one thing, but runtime may be doing something different

**Comment by dfreeman** [comment]
> QUnit doesn't actually throw an assertions error when an assertion fails

Independent of needing the explicit `Assert` annotation, I was already a little hesitant about this change because of that point. Aggregating errors instead of short-circuiting tests on the first failure is one of the differentiating factors between QUnit and most other testing frameworks like Mocha or Jest.

You could argue that `assert.ok(foo)` convey's the user's belief that `foo` should be a truthy value, so it should be ok to allow them to write `foo.bar` on the next line, but that means giving up on QUnit's "run as many assertions as possible" philosophy, since the `TypeError` thrown when you hit `foo.bar` will halt the remainder of the test.

Reasonable people may totally disagree with this opinion, but I'd personally rather have the typechecker force me to write `foo?.bar` in the above scenario, even if it means situations without a one-character workaround (like `assert.ok(typeof baz === 'string')` still leaving `baz: unknown`) require an explicit cast to opt into unsafety for.

> I thought the QUnit Assert type was global? That's how I've been using it. I haven't had to import it? It would still need to be added, yes

This matches my understanding — I believe all of QUnit's types are declared globally, and it's actually `ember-qunit` that sets up `import {...} from 'qunit'` to work. That said, shifting to requiring the annotation still feels like a migration a lot of folks wouldn't be thrilled with.

Overall I _really_ appreciate the work represented by this PR, but I'm not sure the gains outweigh the cost of the change, both in terms of the migration and of the runtime semantics alignment.

File: types/sharedb/index.d.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@
 //                 Eric Hwang <https://github.com/ericyhwang>
 //                 Peter Xu <https://github.com/pxpeterxu>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 3.0
+// TypeScript Version: 3.7
```
Comment (ericyhwang) [inline types/sharedb/index.d.ts:7]: Did you intend to modify the TypeScript Version indicator here?

This PR has no other changes in the `sharedb` types, and I don't believe the types use any language features that require 3.7.

Comment (BarryThePenguin) [reply types/sharedb/index.d.ts:7]: From memory, running `npm test` required it? It looks like the tests for `sharedb` include QUnit as a reference

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2a02c20372ad2dbc512b4ff25864339b9020c259/types/sharedb/sharedb-tests.ts#L1

**Comment by BarryThePenguin** [comment]
Thanks all for your comments 👋 This is exactly the kind of feedback I was after

> Aggregating errors instead of short-circuiting tests on the first failure is one of the differentiating factors between QUnit and most other testing frameworks like Mocha or Jest.

That's a really good point, and not something I was really aware of. My experience has been with Mocha and Jest, so that's what I'm familiar with

> it's actually `ember-qunit` that sets up `import {...} from 'qunit'` to work. That said, shifting to requiring the annotation still feels like a migration a lot of folks wouldn't be thrilled with.

Sounds like if someone wants the short-circuiting behaviour, `ember-mocha` may be a better fit? At least then you have the option of picking your own assertion library 👀 

Again, thanks everyone for your thorough review! I'm happy to close this PR

File: types/sharedb/index.d.ts (hunk 1 above)
Comment (ericyhwang) [reply types/sharedb/index.d.ts:7]: Oh, I somehow missed that when I was looking at the file.

I don't think that file actually uses qunit in any way. May have been a copy-paste by whoever wrote that file originally. The next time someone modifies the file, I'll ask them to remove that reference.

**Pull request closed without merging**